### DISCUSSION TITLE
How to cancel stream when using `ref.listen` and `ref.watch` in `build`

### DISCUSSION THREAD

**Discussion by jibbers42** [rrousselGit/riverpod#1672] [Q&A]
I'd like to cancel the data coming from a long running `StreamProvider` when using `ref.listen` and `ref.watch` in `build`. I'd like to be able to stay on the screen to review the progress shown. Is this possible or do I need to close the screen and open a different one replicating the progress?

I'm currently using a `StreamProvider.autoDispose.family` provider.

**Comment by rrousselGit** [comment]
What do you mean by "cancel the data"?

**Reply by jibbers42** [reply]
I mean cancel the running process. If I have a process running via a StreamProvider, something like below, which I use via `ref.listen` and `ref.watch` in `build`, is there some way to stop the process when the user taps a button, but stay on the screen?

```dart
final streamProviderFamily = StreamProvider.autoDispose.family<int, int>((ref, to) {
  return _genStream(to);
});

Stream<int> _genStream(int to) async* {
  var i = 0;
  while (i < to) {
    await Future<void>.delayed(const Duration(milliseconds: 10));
    i++;
    yield i;
  }
}

class _ProcessState extends ConsumerState<ProcessState > {
  late final _streamProvider = streamProviderFamily (widget.arg);
  var State state = State();

  @override
  Widget build(BuildContext context) {
    ref.listen(
      _streamProvider ,
      (previous, next) {
        next.whenData(_updateStateFromEvent);
      },
    );

    final eventAsync = ref.watch(_streamProvider );

    return eventAsync.when(...);  // shows progress and a Cancel button
  }
}
```

**Reply by rrousselGit** [reply]
No, there is no such thing

**Reply by rrousselGit** [reply]
If you want something like that, maybe make a StateNotifier and manually listen to your stream.
